docs/en/creating-widgets/supplemental.md

```diff
@@ -0,0 +1,972 @@
+
+Dojo optionally supports class-based widgets for applications that prefer the structure of classes over functions. Such widgets inherit from `WidgetBase`, provided by the `@dojo/framework/core/WidgetBase` module, and are required to implement a `render()` method.
+
+The following example shows a trivial yet complete widget within a Dojo application:
+
+> src/widgets/MyWidget.ts
+
+**Function-based Dojo widget variant:**
+
+```ts
+import { create } from '@dojo/framework/core/vdom';
+
+const factory = create();
+
+export default factory(function MyWidget() {
+	return [];
+});
+```
+
+**Class-based Dojo widget variant:**
+
+```ts
+import WidgetBase from '@dojo/framework/core/WidgetBase';
+
+export default class MyWidget extends WidgetBase {
+	protected render() {
+		return [];
+	}
+}
+```
+
+Because this widget returns an empty array from its render function, it has no structural representation within an application's output. Widgets typically [return one or more virtual DOM nodes](#virtual-nodes-example) in order to provide meaningful structure within the application's HTML output.
+
+The process of translating virtual DOM nodes to output on a web page is handled by [Dojo's rendering system](#rendering-in-dojo).
+
+## Widget styling
+
+Styling of a widget's DOM output is handled via CSS, with relevant style classes stored in a CSS module file parallel to the widget's TypeScript module. Styling is identical for both function- and class-based widget variants. This topic is described in detail within the [Styling and Theming reference guide](../styling-and-theming/supplemental.md).
+
+## Scaffolding widgets
+
+Dojo provides the [`dojo create widget`](https://github.com/dojo/cli-create-widget) CLI command to quickly scaffold new widgets. This command helps create:
+
+-   The main TypeScript module for a widget
+-   A unit test for the widget
+-   A CSS module to style the widget
+-   An optional custom element descriptor, if the widget is intended to be used as a Web Component
+
+The command can be installed globally via:
+
+```shell
+npm install -g @dojo/cli-create-widget
+```
+
+and used as:
+
+```shell
+dojo create widget --name <widget name> [--styles <CSS module path>] [--tests <test path>] [--component]
+```
+
+By default, running the command for a given `dojo create widget --name MyScaffoldedWidget` will create the following structure:
+
+    myscaffoldedwidget/
+    ├── MyScaffoldedWidget.ts
+    ├── styles/
+    │   ├── myscaffoldedwidget.m.css
+    │   └── myscaffoldedwidget.m.css.d.ts
+    └── tests/
+        └── unit
+            └── MyScaffoldedWidget.ts
+
+The location where styles and tests are created can be customized using the `--styles` and `--tests` arguments respectively.
+
+A [Custom Element](https://www.w3.org/TR/2016/WD-custom-elements-20161013/) descriptor will be generated if the `--component` argument is passed.
```

Comment:
If we do keep this section, this should be removed as this has been streamlined so that the custom element decorator is not needed to build a custom element.